### PR TITLE
Add support for outputs in the structured renderer

### PR DESCRIPTION
**Pull request by liamcervante**
This change is part of a series of changes that will create a new plan renderer that consumes the JSON structured output produced by the `-json` flag.

This change adds support for rendering and diffing outputs in the structured renderer.

The changes are broken up into a series of PRs that each focus on one part of the renderer to make reviewing easy. This does mean that the order matters, so you should review the PRs in the order listed below. If you prefer to view all the changes in a single (very) big PR, then skip to the last one and change it so it diffs against the main branch instead of the last PR in the chain.

The new renderer has a single entry class: `jsonformat/renderer.go`. It then makes use of two inner packages: `differ` and `change`. The `differ` package converts a single change from the provided JSON structured format into a `Change` object from the `change` package. The `change` package is then responsible for converting the `Change` objects into strings ready to be printed by the external-facing `renderer.go` functions.

This approach differs from the previous renderer where the diff calculation and the rendering happen in the same stage. Separating the diffing and rendering functionality means that going forward adding new kinds of renderers will be easier, and making changes to the format of the output plan will also be simpler.

### PR Chain

| Description         | Link                                                        | Status           |
| ------------------- | ----------------------------------------------------------- | ---------------- |
| Public JSON structs | [#32367](https://github.com/hashicorp/terraform/pull/32367) | Merged         |
| Functions and API   | [#32368](https://github.com/hashicorp/terraform/pull/32368) | Merged          |
| Primitives          | [#32373](https://github.com/hashicorp/terraform/pull/32373) | Merged          |
| Sensitive           | [#32375](https://github.com/hashicorp/terraform/pull/32375) | Merged          |
| Computed            | [#32378](https://github.com/hashicorp/terraform/pull/32378) | Merged          |
| Objects             | [#32391](https://github.com/hashicorp/terraform/pull/32391) | Merged          |
| Replace             | [#32392](https://github.com/hashicorp/terraform/pull/32392) | Merged          |
| Maps                | [#32397](https://github.com/hashicorp/terraform/pull/32397) | Merged          |
| Lists               | [#32401](https://github.com/hashicorp/terraform/pull/32401) | Merged          |
| Sets                | [#32409](https://github.com/hashicorp/terraform/pull/32409) | Merged          |
| Blocks              | [#32422](https://github.com/hashicorp/terraform/pull/32422) | Merged          |
| **Outputs**             | [#32426](https://github.com/hashicorp/terraform/pull/32426) | Merged          |

More PRs are in progress. One will port all the existing unit tests within [format/diff_test.go](https://github.com/hashicorp/terraform/blob/main/internal/command/format/diff_test.go) across to the new renderer to ensure maximum test coverage. I expect that this PR will find many missing edge cases and bugs that need to be fixed. Small bugs will be fixed as part of the main PR, but anything large will be broken out into additional PRs. Another will connect the renderer up to the rest of Terraform. This will be the last PR and it will only be merged once all possible unit tests are passing.

### CODE REVIEW

File: internal/command/jsonformat/change/type_change.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+type typeChangeRenderer struct {
+	NoWarningsRenderer
+
+	before Change
+	after  Change
+}
+
+func (renderer typeChangeRenderer) Render(change Change, indent int, opts RenderOpts) string {
+	opts.overrideNullSuffix = true // Never render null suffix for children of typenm changes.
```
Comment (alisdair) [inline internal/command/jsonformat/change/type_change.go:20]: Is `typenm` a typo?

File: internal/command/jsonformat/differ/output.go [hunk 2]
```diff
@@ -1,7 +1,86 @@
...
+)
+
+const (
+	jsonNumber = "number"
+	jsonObject = "object"
+	jsonArray  = "array"
+	jsonBool   = "bool"
+	jsonString = "string"
+	jsonNull   = "null"
+)
```
Comment (alisdair) [inline internal/command/jsonformat/differ/output.go:18]: This comment applies multiple times through this series of PRs, but I only remembered it here. It'd be nice to have a trivial named type for these string values. We tend to do this elsewhere when we have a fixed enumeration of values, and it's helpful for code navigation and understanding the semantics of function types.

**Review by alisdair — APPROVED** [review]

File: internal/command/jsonformat/differ/output.go (hunk 2 above)
Comment (liamcervante) [reply internal/command/jsonformat/differ/output.go:18]: I've made a note of this, and I'll put together another PR that gets rid of any dangling string types.

File: internal/command/jsonformat/change/type_change.go (hunk 1 above)
Comment (liamcervante) [reply internal/command/jsonformat/change/type_change.go:20]: yes

**Pull request merged**